server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/disk/schemafile/SchemaFile.java

```diff
@@ -236,36 +226,36 @@ public boolean updateStorageGroupNode(IStorageGroupMNode sgNode) throws IOExcept
 
   @Override
   public void writeMNode(IMNode node) throws MetadataException, IOException {
-    int pageIndex;
-    short curSegIdx;
-    ISchemaPage curPage = null;
+    ISchemaPage curPage;
 
-    // Get corresponding page instance, segment id
+    // first segment of the node
     long curSegAddr = getNodeAddress(node);
 
     if (node.isStorageGroup()) {
       curSegAddr = lastSGAddr;
-      pageIndex = getPageIndex(lastSGAddr);
-      curSegIdx = getSegIndex(lastSGAddr);
       isEntity = node.isEntity();
       setNodeAddress(node, lastSGAddr);
     } else {
-      if ((curSegAddr & 0x80000000_00000000L) != 0) {
+      if (curSegAddr < 0L) {
+        // now only 32 bits page index is allowed
         throw new MetadataException(
             String.format(
                 "Cannot store a node with segment address [%s] except for StorageGroupNode.",
                 curSegAddr));
       }
-      pageIndex = SchemaFile.getPageIndex(curSegAddr);
-      curSegIdx = SchemaFile.getSegIndex(curSegAddr);
     }
 
-    // Flush new child
+    long actualAddress; // actual segment to write record
+    IMNode child;
+    ByteBuffer childBuffer;
+    String sk; // search key when split
+    // TODO: reserve order of insert in container may be better
     for (Map.Entry<String, IMNode> entry :
-        ICachedMNodeContainer.getCachedMNodeContainer(node).getNewChildBuffer().entrySet()) {
-
-      // Translate node.child into buffer and Pre-Allocate segment for internal child.
-      IMNode child = entry.getValue();
+        ICachedMNodeContainer.getCachedMNodeContainer(node).getNewChildBuffer().entrySet().stream()
+            .sorted(Map.Entry.<String, IMNode>comparingByKey())
+            .collect(Collectors.toList())) {
+      // check and pre-allocate
+      child = entry.getValue();
```

Comment:
Maybe the ```stream.forEach``` can be directly used here rather than collect as a temp list.

Reply:
Given that it may take thousands of loops, I tried to reuse several variables outside the loop, which makes it difficult to implement with ```stream```. Is it possible to retain as is?